### PR TITLE
REF: extract classes in pandas/core/describe.py

### PR DESCRIPTION
**Pull request by ivanovmg**
- [ ] xref #36833
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Extract classes ``SeriesDescriber`` and ``DataFrameDescriber``.
In the next steps I am going to enable strategy pattern and move corresponding describe functions to the concrete strategy classes (for each datatype).

### CODE REVIEW

File: pandas/core/describe.py [hunk 1]
```diff
@@ -60,107 +61,155 @@ def describe_ndframe(
...
+        obj,
+        include=include,
+        exclude=exclude,
+        datetime_is_numeric=datetime_is_numeric,
+    )
+    result = describer.describe(percentiles=percentiles)
+    return cast(FrameOrSeries, result)
+
 
+def create_describer(
```
Comment (jreback) [inline pandas/core/describe.py:74]: i would just inline this to above

File: pandas/core/describe.py [hunk 2]
```diff
@@ -60,107 +61,155 @@ def describe_ndframe(
...
-    return describe_1d(
-        series,
-        percentiles,
-        datetime_is_numeric,
-        is_series=True,
-    )
 
+    def __init__(
+        self,
+        series: "Series",
```
Comment (jreback) [inline pandas/core/describe.py:125]: can you make this a common method in NDFrame (and changes series -> data) to match dataframe

File: pandas/core/describe.py [hunk 3]
```diff
@@ -60,107 +61,155 @@ def describe_ndframe(
...
+
+class DataFrameDescriber(NDFrameDescriberAbstract):
+    """Class responsible for creating dataframe description.
 
     Parameters
     ----------
-    frame : DataFrame
-        DataFrame to be described.
-    include : 'all', list-like of dtypes or None (default), optional
+    data : DataFrame
```
Comment (jreback) [inline pandas/core/describe.py:146]: match series for data / datetime_is_numeric and add other args

File: pandas/core/describe.py [hunk 4]
```diff
@@ -60,107 +61,155 @@ def describe_ndframe(
...
-        axis=1,
-        sort=False,
-    )
-    d.columns = data.columns.copy()
-    return d
+    def __init__(
+        self,
+        frame: "DataFrame",
+        *,
+        include: Optional[Union[str, Sequence[str]]],
```
Comment (jreback) [inline pandas/core/describe.py:160]: you can do super() for first 2 args

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/describe.py (hunk 2 above)
Comment (ivanovmg) [reply pandas/core/describe.py:125]: I will check that. But I guess it is going to violate Liskov principle.

File: pandas/core/describe.py (hunk 1 above)
Comment (ivanovmg) [reply pandas/core/describe.py]: Done

File: pandas/core/describe.py [hunk 5]
```diff
@@ -376,3 +380,8 @@ def refine_percentiles(percentiles: Optional[Sequence[float]]) -> Sequence[float
         raise ValueError("percentiles cannot contain duplicates")
 
     return unique_pcts
+
+
+def validate_frame(frame: "DataFrame"):
```
Comment (jreback) [inline pandas/core/describe.py:385]: a reason not to do this in the constructor? e.g. L148

Comment (ivanovmg) [reply pandas/core/describe.py:385]: I though of simplifying the constructor.
Ideally I would create a method ``_initialize_frame``, which will perform the validation under the hood and return the frame.

```
self.frame = self._initialize_frame(frame)
```

But that will be a static method, which you prefer not to use.

If you consider that this extraction is unnecessary, then I can definitely inline it into the constructor.

File: pandas/core/describe.py (hunk 2 above)
Comment (ivanovmg) [reply pandas/core/describe.py]: Looks like nothing is violated. It seems to work.

File: pandas/core/describe.py (hunk 5 above)
Comment (ivanovmg) [reply pandas/core/describe.py]: Moved into the constructor.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ivanovmg

**Pull request merged**